File: tests/reference_gpc.c
/*
 * Independent implementation of the Gold Partition certificates.
 *
 * It uses a plain dynamic program on all subsets and recomputes every
 * constrained extension count separately.  It shares no counting code with
 * src/gpc.c.  Input and output follow tests/gpc_classifier_driver.c.
 */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#define NMAX 10

static int order;
static uint16_t above[NMAX];
static uint16_t below[NMAX];
static uint64_t extensions[1 << NMAX];

static uint64_t count_extensions(int first_a, int first_b,
                                 int second_a, int second_b)
{
    const unsigned full = (1u << order) - 1u;
    extensions[0] = 1;
    for (unsigned set = 1; set <= full; set++) {
        uint64_t count = 0;
        for (int x = 0; x < order; x++) {
            if (((set >> x) & 1u) == 0)
                continue;
            const unsigned previous = set & ~(1u << x);
            if (below[x] & ~previous)
                continue;
            if (x == first_b && first_a >= 0 &&
                ((previous >> first_a) & 1u) == 0)
                continue;
            if (x == second_b && second_a >= 0 &&
                ((previous >> second_a) & 1u) == 0)
                continue;
            count += extensions[previous];
        }
        extensions[set] = count;
    }
    return extensions[full];
}

static int slave_count(int x, int y)
{
    const uint16_t full = (uint16_t)((1u << order) - 1u);
    const uint16_t incomparable_x =
        (uint16_t)(full & ~(above[x] | below[x] | (1u << x)));
    const uint16_t incomparable_y =
        (uint16_t)(full & ~(above[y] | below[y] | (1u << y)));
    return __builtin_popcount((unsigned)(above[x] & incomparable_y)) +
           __builtin_popcount((unsigned)(below[y] & incomparable_x));
}

static int read_digraph6(const char *line)
{
    if (line[0] != '&')
        return 0;
    const unsigned char *input = (const unsigned char *)line + 1;
    order = (int)*input++ - 63;
    if (order < 1 || order > NMAX) {
        fprintf(stderr, "unsupported digraph6 order\n");
        exit(2);
    }
    for (int x = 0; x < order; x++)
        above[x] = 0;

    int value = 0;
    int remaining = 0;
    for (int position = 0; position < order * order; position++) {
        if (remaining == 0) {
            if (*input < 63 || *input > 126) {
                fprintf(stderr, "truncated digraph6 record\n");
                exit(2);
            }
            value = (int)*input++ - 63;
            remaining = 6;
        }
        remaining--;
        if ((value >> remaining) & 1) {
            const int x = position / order;
            const int y = position % order;
            above[x] |= (uint16_t)(1u << y);
        }
    }

    for (int pass = 0; pass < order; pass++)
        for (int x = 0; x < order; x++)
            for (int y = 0; y < order; y++)
                if ((above[x] >> y) & 1u)
                    above[x] |= above[y];
    for (int x = 0; x < order; x++)
        below[x] = 0;
    for (int x = 0; x < order; x++)
        for (int y = 0; y < order; y++)
            if ((above[x] >> y) & 1u)
                below[y] |= (uint16_t)(1u << x);
    return 1;
}

static char classify(void)
{
    int incomparable = 0;
    for (int x = 0; x < order; x++)
        for (int y = x + 1; y < order; y++)
            if (((above[x] >> y) & 1u) == 0 &&
                ((above[y] >> x) & 1u) == 0)
                incomparable = 1;
    if (!incomparable)
        return 'C';

    const uint64_t total = count_extensions(-1, -1, -1, -1);
    uint64_t pair[NMAX][NMAX] = {{0}};
    for (int x = 0; x < order; x++)
        for (int y = 0; y < order; y++) {
            if (x == y)
                continue;
            if ((above[x] >> y) & 1u)
                pair[x][y] = total;
            else if ((above[y] >> x) & 1u)
                pair[x][y] = 0;
            else if (x < y)
                pair[x][y] = count_extensions(x, y, -1, -1);
            else
                pair[x][y] = total - pair[y][x];
        }

    for (int x = 0; x < order; x++)
        for (int y = 0; y < order; y++) {
            if (x == y || ((above[x] >> y) & 1u) ||
                ((above[y] >> x) & 1u))
                continue;
            if (2 * pair[x][y] == total ||
                (2 * pair[x][y] >= total && slave_count(x, y) <= 1))
                return 'P';
        }

    for (int x = 0; x < order; x++)
        for (int y = 0; y < order; y++) {
            if (x == y || 2 * pair[y][x] > total)
                continue;
            for (int z = 0; z < order; z++) {
                if (z == x || z == y || 2 * pair[z][y] > total)
                    continue;
                const uint64_t bound =
                    pair[y][x] > pair[z][y] ? pair[y][x] : pair[z][y];
                if (count_extensions(x, y, y, z) <= bound)
                    return 'T';
            }
        }
    return 'O';
}

int main(void)
{
    char line[4096];
    while (fgets(line, sizeof(line), stdin))
        if (read_digraph6(line))
            puts((char[2]){classify(), '\0'});
    return 0;
}

File: tests/reference_gpc.py
#!/usr/bin/env python3
"""Independent small-order check of the Gold Partition certificates.

This program neither calls nor translates the C implementation.  It generates
unlabeled posets by adjoining a maximal element over every order ideal, removes
isomorphic duplicates by canonical relabeling, enumerates linear extensions,
and tests the three sufficient conditions in Peczarski's formulation.
"""

from __future__ import annotations

from itertools import permutations
import sys


A000112 = {1: 1, 2: 2, 3: 5, 4: 16, 5: 63, 6: 318, 7: 2_045}


def bits(mask: int):
    while mask:
        bit = mask & -mask
        yield bit.bit_length() - 1
        mask ^= bit


def downsets_from_up(up: tuple[int, ...]) -> list[int]:
    down = [0] * len(up)
    for x, above in enumerate(up):
        for y in bits(above):
            down[y] |= 1 << x
    return down


def invariant_colors(up: tuple[int, ...]) -> list[int]:
    down = downsets_from_up(up)
    keys = [(up[x].bit_count(), down[x].bit_count()) for x in range(len(up))]
    while True:
        distinct = {key: rank for rank, key in enumerate(sorted(set(keys)))}
        colors = [distinct[key] for key in keys]
        refined = [
            (
                colors[x],
                tuple(sorted(colors[y] for y in bits(up[x]))),
                tuple(sorted(colors[y] for y in bits(down[x]))),
            )
            for x in range(len(up))
        ]
        if refined == keys:
            return colors
        keys = refined


def canonical(up: tuple[int, ...]) -> tuple[int, ...]:
    n = len(up)
    colors = invariant_colors(up)
    blocks: dict[int, list[int]] = {}
    for vertex, color in enumerate(colors):
        blocks.setdefault(color, []).append(vertex)
    ordered_blocks = [blocks[color] for color in sorted(blocks)]
    relabel = [0] * n
    best: tuple[int, ...] | None = None

    def search(block_index: int, next_label: int) -> None:
        nonlocal best
        if block_index == len(ordered_blocks):
            candidate = [0] * n
            for x in range(n):
                candidate[relabel[x]] = sum(
                    1 << relabel[y] for y in bits(up[x])
                )
            encoded = tuple(candidate)
            if best is None or encoded < best:
                best = encoded
            return
        block = ordered_blocks[block_index]
        for ordering in permutations(block):
            for offset, vertex in enumerate(ordering):
                relabel[vertex] = next_label + offset
            search(block_index + 1, next_label + len(block))

    search(0, 0)
    assert best is not None
    return best


def order_ideals(up: tuple[int, ...]) -> list[int]:
    down = downsets_from_up(up)
    return [
        subset
        for subset in range(1 << len(up))
        if all(not (down[x] & ~subset) for x in bits(subset))
    ]


def generate_posets(maximum_order: int) -> dict[int, list[tuple[int, ...]]]:
    result = {1: [(0,)]}
    for n in range(2, maximum_order + 1):
        seen: set[tuple[int, ...]] = set()
        for smaller in result[n - 1]:
            for ideal in order_ideals(smaller):
                extended = list(smaller) + [0]
                for x in bits(ideal):
                    extended[x] |= 1 << (n - 1)
                seen.add(canonical(tuple(extended)))
        result[n] = sorted(seen)
        if len(result[n]) != A000112[n]:
            raise AssertionError(
                f"order {n}: generated {len(result[n])}, "
                f"expected {A000112[n]}"
            )
    return result


def linear_extensions(up: tuple[int, ...]) -> list[tuple[int, ...]]:
    n = len(up)
    down = downsets_from_up(up)
    extensions: list[tuple[int, ...]] = []

    def visit(prefix: tuple[int, ...], used: int) -> None:
        if len(prefix) == n:
            extensions.append(prefix)
            return
        for x in range(n):
            if not (used >> x) & 1 and not (down[x] & ~used):
                visit(prefix + (x,), used | (1 << x))

    visit((), 0)
    return extensions


def slave_count(
    up: tuple[int, ...], down: list[int], x: int, y: int
) -> int:
    n = len(up)
    full = (1 << n) - 1
    incomparable_y = full & ~(up[y] | down[y] | (1 << y))
    incomparable_x = full & ~(up[x] | down[x] | (1 << x))
    return (up[x] & incomparable_y).bit_count() + (
        down[y] & incomparable_x
    ).bit_count()


def has_certificate(up: tuple[int, ...]) -> bool:
    n = len(up)
    down = downsets_from_up(up)
    incomparable = [
        (x, y)
        for x in range(n)
        for y in range(x + 1, n)
        if not ((up[x] >> y) & 1) and not ((up[y] >> x) & 1)
    ]
    if not incomparable:
        return True

    for x, y in incomparable:
        if (
            slave_count(up, down, x, y) <= 1
            and slave_count(up, down, y, x) <= 1
        ):
            return True

    extensions = linear_extensions(up)
    positions = [
        {vertex: index for index, vertex in enumerate(extension)}
        for extension in extensions
    ]
    total = len(extensions)

    def before(x: int, y: int) -> int:
        return sum(position[x] < position[y] for position in positions)

    for x, y in incomparable:
        xy = before(x, y)
        if 2 * xy == total:
            return True
        if 2 * xy >= total and slave_count(up, down, x, y) <= 1:
            return True
        if 2 * (total - xy) >= total and slave_count(up, down, y, x) <= 1:
            return True

    for x in range(n):
        for y in range(n):
            if y == x:
                continue
            yx = before(y, x)
            if 2 * yx > total:
                continue
            for z in range(n):
                if z == x or z == y:
                    continue
                zy = before(z, y)
                bound = max(yx, zy)
                if 2 * bound > total:
                    continue
                xyz = sum(
                    position[x] < position[y] < position[z]
                    for position in positions
                )
                if xyz <= bound:
                    return True
    return False


def main() -> None:
    maximum_order = int(sys.argv[1]) if len(sys.argv) == 2 else 7
    if maximum_order not in A000112:
        raise SystemExit("maximum order must lie between 1 and 7")
    posets = generate_posets(maximum_order)
    for n in range(1, maximum_order + 1):
        chains = 0
        open_classes = 0
        for up in posets[n]:
            incomparable = sum(
                not ((up[x] >> y) & 1) and not ((up[y] >> x) & 1)
                for x in range(n)
                for y in range(x + 1, n)
            )
            chains += incomparable == 0
            open_classes += not has_certificate(up)
        if chains != 1 or open_classes:
            raise SystemExit(
                f"order {n}: chains={chains}, open={open_classes}"
            )
        print(
            f"REFERENCE n={n} total={len(posets[n])} "
            f"chain={chains} open={open_classes}"
        )
    print("PASS: independent small-order certificate check")


if __name__ == "__main__":
    main()
